src/Swarm/Util/WindowedCounter.hs

```diff
@@ -0,0 +1,179 @@
+import Prelude hiding (length)
+
+-- | Values that can be offset by an integral amount
+class Offsettable a where
+  offsetBy :: Int -> a -> a
+
+-- | A "sliding window" of a designated span that supports insertion
+-- of tick "timestamps" that represent some state of interest during that tick.
+-- This data structure supports efficient querying of the ratio of
+-- {ticks for which that state existed}
+-- to the
+-- {total number of ticks spanned by the window}.
+--
+-- The primary use case is in displaying the "activity level" of a robot.
+--
+-- == Efficiency considerations
+--
+-- The data retention of the window shall be maintained externally by
+-- invoking the 'discardGarbage' function. However, we should not
+-- unconditionally invoke this function upon each game tick.
+--
+-- For efficiency, we do not want to iterate over every robot
+-- upon every tick; we only want to "visit" a robot if it is actually
+-- doing work that tick.
+-- Because of this, there may be some ticks in which the oldest element
+-- that is still stored falls outside of the nominal retention window
+-- while a robot is inactive.
+--
+-- One might think we could perform garbage collection whenever we execute queries.
+-- However, in the context in which the view powered by the query is generated, we
+-- are not permitted to mutate the "state" of the game
+-- (the type signature of the rendering function is @AppState -> Widget Name@).
+--
+-- Therefore, when we perform "queries" on the window, we must apply some
+-- filtering to exclude the "stragglers"; data members that have already fallen outside
+-- the window but have not yet been "garbage collected".
+-- We use a 'Set' to allow this filtering to be performed in @O(log n)@ time.
+--
+-- In the worst case, the entire dataset may "age out" without being garbage collected,
+-- so that an @O(log n)@ filtering operation might be performed upon every "frame refresh"
+-- of the UI view.
+-- However, we also store the largest element of the window separately from the 'Set' so that
+-- we can compare against it for a @O(1)@ short-circuited path once every member ages out.
+--
+-- The maximum number of elements ever stored in the 'Set' will be the width of the nominal
+-- span, even after some protracted failure to "garbage collect".
+data WindowedCounter a = WindowedCounter
+  { _members :: Set a
+  , _lastLargest :: Maybe a
+  -- ^ NOTE: It is possible that '_lastLargest' may not exist in the 'Set'.
+  , _nominalSpan :: Int
+  -- ^ Data retention window. This value is guaranteed positive by the smart constructor.
+  }
+  deriving (Eq, Show)
+
+-- Automatically deriving 'FromJSON' circumvents the protection offered by "smart constructors",
+-- and the 'ToJSON' instance may expose internal details.
+-- Therefore, we write our own custom implementations.
+instance ToJSON (WindowedCounter a) where
+  toJSON (WindowedCounter membersSet _lastLargest nominalSpan) =
+    object
+      [ "members" .= Set.size membersSet
```

Comment:
> Why do you write a `members` field but then not read it in the `FromJSON` instance?

Added a comment.  IMO, any creation of internal state should have to go through the smart constructor, as arbitrary JSON does not necessarily uphold the same invariants.

> Also, what's the justification for throwing away information about the members of the set and only writing its size?

At first, I was thinking that we shouldn't "reveal" information that is supposed to be "encapsulated".  But, since the `ToJSON` instance is supposed to be for diagnostics, I now expose the entire `Set`.

Reply:
Ah, OK, that makes sense.  And it won't really matter if the `WindowedCounter`s are not preserved when saving + loading a game---it just means it will take a little bit for the activity% values for the robots to "warm up" again.